pkg/controllers/cisnetworkpolicy/controller.go

```diff
@@ -0,0 +1,166 @@
+	coreclient "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+	core "k8s.io/api/core/v1"
+	netv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	FlannelPresenceLabel         = "flannel.alpha.coreos.com/public-ip"
+	FlannelHostNetworkPolicyName = "rke2-flannel-host-networking"
+)
+
+func CISNetworkPolicyController(ctx context.Context, sc *server.Context) error {
+	Register(ctx, sc.Core.Core().V1().Node(), sc.K8s)
+
+	return nil
+}
+
+func Register(ctx context.Context,
+	nodes coreclient.NodeController,
+	k8s kubernetes.Interface,
+) error {
+	h := &handler{
+		ctx: ctx,
+		k8s: k8s,
+	}
+	logrus.Debugf("CISNetworkPolicyController: Registering controller hooks")
+	nodes.OnChange(ctx, "cisnetworkpolicy-node", h.onChange)
+	nodes.OnRemove(ctx, "cisnetworkpolicy-node", h.onRemove)
+	return nil
+}
+
+type handler struct {
+	ctx context.Context
+	k8s kubernetes.Interface
+}
+
+func (h *handler) onRemove(key string, node *core.Node) (*core.Node, error) {
+	return h.reconcileFlannelHostNetworkPolicy(key, node)
+}
+
+func (h *handler) onChange(key string, node *core.Node) (*core.Node, error) {
+	if node == nil {
+		return nil, nil
+	}
+
+	return h.reconcileFlannelHostNetworkPolicy(key, node)
+}
+
+func (h *handler) reconcileFlannelHostNetworkPolicy(nodeName string, node *core.Node) (*core.Node, error) {
+	var np *netv1.NetworkPolicy
+	var npIR *netv1.NetworkPolicyIngressRule
+
+	npIR, err := h.generateHostNetworkPolicyIngressRule()
+	if err != nil {
+		return nil, err
+	}
+
+	np, err = h.generateHostNetworkingNetworkPolicy(npIR)
+	if err != nil {
+		return nil, err
+	}
+
+	var namespaces = []string{
+		metav1.NamespaceSystem,
+		metav1.NamespaceDefault,
+		metav1.NamespacePublic,
+	}
+
+	for _, namespace := range namespaces {
+		applied := false
+		for !applied {
+			if npGet, err := h.k8s.NetworkingV1().NetworkPolicies(namespace).Get(h.ctx, FlannelHostNetworkPolicyName, metav1.GetOptions{}); err != nil {
+				if errors.IsNotFound(err) {
+					if _, err := h.k8s.NetworkingV1().NetworkPolicies(namespace).Create(h.ctx, np, metav1.CreateOptions{}); err != nil {
+						if !errors.IsAlreadyExists(err) {
+							logrus.Errorf("CISNetworkPolicyController: Error creating network policy in namespace %s: %v", namespace, err)
+							return nil, err
+						}
+					}
+				} else {
+					logrus.Errorf("CISNetworkPolicyController: Error getting network policy network policy in namespace %s: %v", namespace, err)
+					return nil, err
+				}
+			} else {
+				npGet = npGet.DeepCopy()
+				npGet.Spec.Ingress[0] = *npIR
+				_, err := h.k8s.NetworkingV1().NetworkPolicies(namespace).Update(h.ctx, npGet, metav1.UpdateOptions{})
+				if err != nil {
+					if errors.IsConflict(err) {
+						logrus.Errorf("CISNetworkPolicyController: Conflict updating network policy network policy in namespace %s: %v", namespace, err)
+						continue
```

Comment:
Should this use a retry-with-backoff strategy instead of just tightly looping until it succeeds? I think there are some helpers for doing that.

Reply:
Now using retry